EmbeddedPkg/Library/NonCoherentDmaLib/NonCoherentDmaLib.c

```diff
@@ -553,11 +553,11 @@ DmaAllocateAlignedBuffer (
 
   InsertHeadList (&UncachedAllocationList, &Alloc->Link);
 
-  // Remap the region with the new attributes
+  // Remap the region with the new attributes and mark it non-executable
   Status = gDS->SetMemorySpaceAttributes (
                   (PHYSICAL_ADDRESS)(UINTN)Allocation,
                   EFI_PAGES_TO_SIZE (Pages),
-                  MemType
+                  MemType | EFI_MEMORY_XP

```

Comment:
I ran into an issue with this on Raspberry Pi 4: that platform does not assign the write/execute protectable GCD capabilities, and so this call will always fail. We can fix that, of course, but it would be nice if this would handle that case a bit better.

Reply:
@ardbiesheuvel, thanks for letting me know. I can put up a PR to set the capabilities before the attributes (which is the correct pattern).

On the x86 side, CpuDxe sets the EFI_MEMORY_RP, EFI_MEMORY_XP, and EFI_MEMORY_RO capabilities for all memory regions to avoid issues like this (and actually it mostly helps in the GCD sync). I had put up a patch to do the same on the ARM64 side maybe a year ago or so and we had some discussion on it, tbh I forget where that ended but it did not get merged. As part of the current Project Mu upstreaming push we are doing, I am planning to reintroduce that commit, we can chat about its merits. If you want a preview, here is what we do on the Mu side: https://github.com/microsoft/mu_silicon_arm_tiano/commit/757c8a936bbb6060607e3aba2f348e1820eda2c6. Here is the equivalent x86 edk2 code: https://github.com/tianocore/edk2/blob/be1d408773504df8b95a2a81be7b558276ba5356/UefiCpuPkg/CpuDxe/CpuPageTable.c#L994-L1016.

Reply:
@ardbiesheuvel, put up a PR for the EmbeddedPkg issue here: https://github.com/tianocore/edk2/pull/5999.